backend/src/services/hsm/hsm-service.ts

```diff
@@ -0,0 +1,349 @@
+    const unwrapTemplate = {
+      class: graphene.ObjectClass.SECRET_KEY,
+      keyType: graphene.KeyType.AES
+    } as grapheneLib.ITemplate;
+
+    const unwrappingMech = {
+      name: "AES_KEY_WRAP",
+      params: null
+    } as grapheneLib.MechanismType;
+
+    return session.unwrapKey(unwrappingMech, new graphene.Key(masterKey).toType(), wrappedKey, unwrapTemplate);
+  };
+
+  const $keyExists = (session: grapheneLib.Session): boolean => {
+    try {
+      const key = $findMasterKey(session);
+      // items(0) will throw an error if no items are found
+      // Return true only if we got a valid object with handle
+      return key && typeof key.handle !== "undefined";
+    } catch (error) {
+      // If items(0) throws, it means no key was found
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-call
+      if ((error as any).message?.includes("CKR_OBJECT_HANDLE_INVALID")) {
+        return false;
+      }
+      logger.error(error, "Error checking for HSM key presence");
+      return false;
+    }
+  };
+
+  const encrypt: {
+    (data: Buffer, providedSession: grapheneLib.Session): Promise<Buffer>;
+    (data: Buffer): Promise<Buffer>;
+  } = async (data: Buffer, providedSession?: grapheneLib.Session) => {
+    if (!module) {
+      throw new Error("PKCS#11 module is not initialized");
+    }
+
+    const $performEncryption = (s: grapheneLib.Session) => {
+      // Generate IV for encryption
+      const iv = s.generateRandom(IV_LENGTH);
+
+      // Generate and wrap a new session key
+      const { wrappedKey, sessionKey } = $generateAndWrapKey(s);
+
+      const alg = {
+        name: appCfg.HSM_MECHANISM,
+        params: new graphene.AesGcm240Params(iv)
+      } as grapheneLib.IAlgorithm;
+
+      const cipher = s.createCipher(alg, new graphene.Key(sessionKey).toType());
+
+      // Calculate the output buffer size based on input length
+      // GCM adds a 16-byte auth tag, so we need input length + 16
+      const outputBuffer = Buffer.alloc(data.length + TAG_LENGTH);
+      const encryptedData = cipher.once(data, outputBuffer);
+
+      // Format: [Wrapped Key (40)][IV (16)][Encrypted Data + Tag]
+      return Buffer.concat([wrappedKey, iv, encryptedData]);
+    };
+
+    if (providedSession) {
+      return $performEncryption(providedSession);
+    }
+
+    const encrypted = await withSession({ module, graphene }, $performEncryption);
+
+    return encrypted;
+  };
+
+  const decrypt: {
+    (encryptedBlob: Buffer, providedSession: grapheneLib.Session): Promise<Buffer>;
+    (encryptedBlob: Buffer): Promise<Buffer>;
+  } = async (encryptedBlob: Buffer, providedSession?: grapheneLib.Session) => {
+    if (!module) {
+      throw new Error("HSM service not initialized");
+    }
+
+    const $performDecryption = (s: grapheneLib.Session) => {
+      const wrappedKey = encryptedBlob.subarray(0, WRAPPED_KEY_LENGTH);
```

Comment:
Just to clarify my understanding, is `wrappedKey` included together in `encryptedBlob` because they are stored together in the database? (i.e. when encrypting root key with HSM then the resulting ciphertext contains the encrypted root key, IV, and wrapped key so that when decrypting it we unwrap it with the master key in HSM and use that to decrypt the root key).

Is that correct? 

Also I'm curious about the need for this wrapping key — Can you point me to some resources discussing this as a best practice etc.?

Reply:
Your understand is correct. To unwrap the wrapped key, you need the HSM master key. This is to limit exposure / usage of the master HSM key as much as possible.

https://cryptography.io/en/latest/hazmat/primitives/keywrap/
https://security.stackexchange.com/questions/270194/what-is-the-point-of-the-aes-key-wrap-algorithm

You can also think of the wrapping mechanism as a form of envelope encryption.